File: rob.java
/ Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach:
// Use two variables: skip for the maximum money if you don't rob the current house, and take for the maximum money if you do rob it.
// For each house, update skip to the max value of skip or take from the previous house, and update take to skip from
// the previous house plus the current house's money. After processing all houses, the maximum money you can rob is the higher value between skip and take.

class Solution {
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0){
            return -1;
        }

        int skip = 0;
        int take = nums[0];

        for(int i = 1; i < nums.length; i++){
            int temp = skip;
            skip = Math.max(temp, take);
            take = temp + nums[i];
        }
        return Math.max(skip, take);
    }
}

File: rob.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        rob1, rob2 = 0, 0

        for n in nums:
            temp = max(n + rob1, rob2)
            rob1 = rob2
            rob2 = temp

        return rob2  
# time complexity: O(n) because we only need to iterate through the nums array once
# space complexity: O(1) because we are only using a constant amount of extra space
